File: examples/typescript/sdk-reference/error-handling/basic-error-handling.ts
import { withDurableExecution, StepError } from "@aws/durable-execution-sdk-js";

export const handler = withDurableExecution(async (event, context) => {
  try {
    const result = await context.step("process-order", async (stepCtx) => {
      const orderId = (event as { orderId?: string }).orderId;
      if (!orderId) {
        throw new Error("orderId is required");
      }
      return { orderId, status: "processed" };
    });
    return result;
  } catch (err) {
    if (err instanceof StepError) {
      context.logger.error("Step failed", { cause: err.cause?.message });
      return { error: err.cause?.message ?? "Step failed" };
    }
    throw err;
  }
});

File: examples/csharp/sdk-reference/error-handling/basic-error-handling.cs
using Amazon.Lambda.Core;
using Amazon.Lambda.DurableExecution;
using Microsoft.Extensions.Logging;

public class BasicErrorHandlingExample
{
    public Task<DurableExecutionInvocationOutput> Handler(
        DurableExecutionInvocationInput input, ILambdaContext context)
        => DurableFunction.WrapAsync<OrderEvent, OrderResult>(Workflow, input, context);

    private async Task<OrderResult> Workflow(OrderEvent input, IDurableContext ctx)
    {
        try
        {
            OrderResult result = await ctx.StepAsync(
                async (_, _) =>
                {
                    if (string.IsNullOrEmpty(input.OrderId))
                    {
                        throw new InvalidOperationException("orderId is required");
                    }
                    return new OrderResult(input.OrderId, Status: "processed");
                },
                name: "process-order");
            return result;
        }
        catch (StepException e)
        {
            // The step exhausted its retries; the SDK checkpointed the final
            // error and threw it here.
            ctx.Logger.LogError("Step failed: {Message}", e.Message);
            return new OrderResult(input.OrderId, Status: "error", Error: e.Message);
        }
    }
}

public record OrderEvent(string OrderId);
public record OrderResult(string OrderId, string Status, string? Error = null);
